artiq/master/worker.py

```diff
@@ -157,8 +158,16 @@ async def close(self, term_timeout=2.0):
 
     async def _send(self, obj, cancellable=True):
         assert self.io_lock.locked()
-        line = pyon.encode(obj)
-        self.ipc.write((line + "\n").encode())
+
+        data = pickle.dumps(obj)
+        size_str = (str(len(data)) + "\n").encode()
+
+        # Write the size of the data, delimited by \n. Then, write the data
+        # itself. This might contain \n chars, but the receiving reader knows
+        # how many bytes it's expecting so it doesn't need delimiters
```

Comment:
Why not move everything to a binary protocol with length (with struct.pack)+binary data and no special meaning for \n?

Reply:
Could do, but the length needs delimiting in some way and newline delimiting is already implemented and well tested.

Reply:
The loading of the binary data does completely ignore `\n`, it's only relevant for the length data.